crates/usvg/src/lib.rs

```diff
@@ -69,6 +81,125 @@ pub trait TreeWriting {
 
 impl TreeWriting for usvg_tree::Tree {
     fn to_string(&self, opt: &XmlOptions) -> String {
-        writer::convert(self, opt)
+        let used_ids = get_ids(self);
+        let mut id_generator = IdGenerator::new(used_ids);
+
+        let mut writer_context = WriterContext {
+            tree: &self,
+            opt,
+            text_path_map: Vec::new(),
+            id_generator: &mut id_generator,
+        };
+
+        writer::convert(&mut writer_context)
+    }
+}
+
+/// A generator for new IDs.
+#[derive(Default)]
+pub(crate) struct IdGenerator {
+    /// A set of IDs that should never be generated by the generator.
+    ids: HashSet<String>,
+    /// The counter for the current path ID.
+    path: u64,
+}
+
+impl IdGenerator {
+    pub fn new(ids: HashSet<String>) -> Self {
+        Self {
+            ids,
+            ..Self::default()
+        }
+    }
+
+    pub fn bump_path(&mut self) -> String {
+        IdGenerator::bump_impl(&self.ids, &mut self.path, "p")
+    }
+
+    fn bump_impl(ids: &HashSet<String>, field: &mut u64, format_str: &str) -> String {
+        let mut bump = || {
+            *field += 1;
+            let result = format!("{}{}", format_str, field);
+            result
+        };
+
+        let mut id = bump();
+
+        while ids.contains(&id) {
+            id = bump();
+        }
+
+        id
+    }
+}
+
+fn get_ids(tree: &Tree) -> HashSet<String> {
+    let mut ids = HashSet::new();
+    get_ids_impl(&tree.root, &mut ids);
+    ids.remove("");
+    ids
+}
+
+fn get_ids_impl(node: &Node, ids: &mut HashSet<String>) {
```

Comment:
Ugh... I guess I would have to revert to the old tree representation, the one with defs. This way there would be no need for such ugly code.
But that approach has it's own downsides as well.